File: Oxygene/Toffee/macOS/NSTableView/NSTableView/MainWindowController.pas
namespace NSTableView;

interface

uses
  AppKit,
  Foundation;

type
  [IBObject]
  MainWindowController = public class(NSWindowController, INSTableViewDelegate,INSTableViewDataSource)
  private
    people : NSMutableArray;
  protected
  public
    [IBOutlet] tableView : NSTableView;
    [IBOutlet] messageLabel : NSTextField;

    method init: id; override;

    [IBAction] method addPerson(sender : id);
    [IBAction] method removePerson(sender : id);
    method updateMessageLabel();
   

    //INSTableViewDataSource
    method tableView(tblView : NSTableView) objectValueForTableColumn(tableColumn: NSTableColumn) row(row : NSInteger): id;
    method tableView(tblView : NSTableView) setObjectValue(obj: id) forTableColumn(tableColumn : NSTableColumn) row(row: NSInteger);
    method numberOfRowsInTableView(tblView : NSTableView): NSInteger;

    //iNSTableViewDelegate
    method tableViewSelectionDidChange(aNotification : NSNotification);
  end;

implementation

method MainWindowController.init: id;
begin
  self := inherited initWithWindowNibName('MainWindowController');
  if assigned(self) then begin
    self.people := new NSMutableArray();
  end;
  result := self;
end;


method MainWindowController.addPerson(sender: id);
begin
   var person := new Person();
   person.name :=  "New Person";
   person.age := 21;
   self.people.addObject(person);
   self.tableView.reloadData();
   self.updateMessageLabel();
end;

method MainWindowController.removePerson(sender: id);
begin
   var row := self.tableView.selectedRow();
   if (row > -1) and (row < self.people.count()) then begin
     self.tableView.abortEditing();
     self.people.removeObjectAtIndex(row);
     self.tableView.reloadData();
     self.updateMessageLabel();
   end;
end;

method MainWindowController.updateMessageLabel();
begin
   var row := self.tableView.selectedRow();
   if (row > -1) and (row < self.people.count()) then begin
     var person := self.people.objectAtIndex(row);
     self.messageLabel.stringValue := person.name;
   end else begin
     self.messageLabel.stringValue := "";
   end;
end;

method MainWindowController.tableView(tblView: NSTableView) objectValueForTableColumn(tableColumn: NSTableColumn) row(row: NSInteger): id;
begin
  var person := self.people.objectAtIndex(row);
  case tableColumn.identifier of 
    "name": 
      result := person.name;
    "age": 
      result := person.age;
    else 
      result := "Unknown";
  end;
end;

method MainWindowController.tableView(tblView: NSTableView) setObjectValue(obj: id) forTableColumn(tableColumn: NSTableColumn) row(row: NSInteger);
begin
  var person : Person := self.people.objectAtIndex(row);
  case tableColumn.identifier of 
    "name": 
      person.name := obj;
    "age": 
      person.age := obj.intValue;
  end;
  self.updateMessageLabel();
end;

method MainWindowController.numberOfRowsInTableView(tblView: NSTableView): NSInteger;
begin
  result := self.people.count();
end;

method MainWindowController.tableViewSelectionDidChange(aNotification: NSNotification);
begin
  self.updateMessageLabel();
end;

end.

File: Hydrogene/Toffee/macOS/NSTableView/NSTableView/MainWindowController.cs
using AppKit;

namespace NSTableView
{
	[IBObject]
	public class MainWindowController : NSWindowController, INSTableViewDataSource, INSTableViewDelegate
	{

		[IBOutlet]
		public NSTableView tableView;

		[IBOutlet]
		public NSTextField messageLabel;

		private NSMutableArray people;


		public override id init()
		{
			this = base.initWithWindowNibName("MainWindowController");
			if (this != null)
			{
				this.people = new NSMutableArray();
			}
			return this;
		}

		public override void windowDidLoad()
		{
			base.windowDidLoad();

			// Implement this method to handle any initialization after your window controller's
			// window has been loaded from its nib file.
		}

		[IBAction]
		public void addPerson(id  sender)
		{
			Person person = new Person();
			person.name = "New Person";
			person.age = 21;
			this.people.addObject(person);
			this.tableView.reloadData();
			self.updateMessageLabel();  
		}

		[IBAction]
		public void removePerson(id  sender)
		{
			this.tableView.abortEditing();
			int row = this.tableView.selectedRow();
			if (row > -1 && row < this.people.count())
			{
				this.people.removeObjectAtIndex(row);
				this.tableView.reloadData();
				self.updateMessageLabel();
			}
		}

		public void updateMessageLabel()
		{
			int row = this.tableView.selectedRow();
			if (row > -1 && row < this.people.count()) {
				Person person = this.people.objectAtIndex(row);
				this.messageLabel.stringValue = person.name;
			}
			else
			{
				this.messageLabel.stringValue = "";
			}
		}
	  
		#region ITableViewDataSource

		public NSInteger numberOfRowsInTableView(NSTableView tableView)
		{
			return this.people.count();
		}

		public id tableView(NSTableView tableView) objectValueForTableColumn(NSTableColumn tableColumn) row(NSInteger row)
		{
			Person person = this.people.objectAtIndex(row);
			if (tableColumn.identifier == "name")
			{
				return person.name;
			}
			else if (tableColumn.identifier == "age")
			{
				return person.age;
			} else {
				return "Unkown";
			}
		}
	
		public void tableView(NSTableView tableView) setObjectValue(id obj) forTableColumn(NSTableColumn tableColumn) row(NSInteger row)
		{
			Person person = this.people.objectAtIndex(row);
			if (tableColumn.identifier == "name")
			{
				person.name = obj;
			}
			else if (tableColumn.identifier == "age")
			{
				person.age = obj.intValue;
			}
			self.updateMessageLabel();
		}

		#endregion

		#region ITableViewDelegate

		public void tableViewSelectionDidChange(NSNotification aNotification)
		{
			self.updateMessageLabel();
		}

		#endregion

	}
}
